### ISSUE TITLE
PubSubReactiveFactory keeps messages in flight on shutdown

### ISSUE THREAD

**Issue by jgordijn** [GoogleCloudPlatform/spring-cloud-gcp#649]
Current implementation of the `PubSubReactiveFactory` does a `pullAsync` in the `backpressurePull`. Problem is that this call is async and when the pipeline is shutdown there may still be a call in flight. When this call completes and the callback notices that the `sink.isCancelled`, the payload is ignored. This leads to messages hitting the deadline of the ack period when restarting (scaling down, redeploy, etc) the application. Using auto scaling this happens a lot to us.

BTW the pipeline is shutdown with a flag which is set upon spring shutdown pre-notification, instead of calling the `dispose()` function. This to make sure that we would handle all in flight messages before shutting down, as opposed to cancelling outstanding jobs.

```kotlin
val shutdown = AtomicBoolean(false)
pubSubReactiveFactory.poll("topic", 100)
            .takeUntil { shutdown.get() }
            .flatMap { ... }
            .subscribeOn(Schedulers.parallel())
            .subscribe()
```

**Comment by elefeint** [comment]
That's an interesting scenario. On a coarse level, everything works as expected in each layer -- Cloud Pub/Sub re-delivers the messages after ackDeadline expires. But for your requirements it creates an inconvenience.

It would definitely make sense to `nack()` all pubsub messages that come back from `pullAsync()` if `sink.isCancelled()` is detected (which it should once `takeUntil` detects the exit condition). Would this help with your scenario to have the messages redelivered immediately instead of waiting for ackDeadline expiration?

**Comment by jgordijn** [comment]
Yes, that is exactly what I was thinking about. It is a best effort and I think in our case we should also lower the ack deadline, but nacking the unnecessary frame seems like a nice improvement.
